File: BOJ_Swift/BOJ_2517_달리기_P4.swift
//
//  SwiftAlgorithm
//
//  Created by zun on 09/02/2020.
//  Copyright © 2020 zun. All rights reserved.
//

// 순위와 실력을 저장할 구조체
struct Player {
  var index: Int
  var power: Int
}

let n = readLine()!.split(separator: " ").compactMap{ Int(String($0)) }[0]

// 정보를 저장할 배열
var arr = [Player](repeating: Player(index: 0, power: 0), count: 500001)
// merge sort를 구현하기 위해 사용될 임시 배열
var temp = [Player](repeating: Player(index: 0, power: 0), count: 500001)
// 답을 출력할 배열. index 별로 현재 달리고 있는 선수의 순위의 최선의 등수를 저장
var ans = [Int](repeating: 0, count: 500001)

func mergeSort(left: Int, right: Int) {
  if left == right {
    return
  } else {
    let mid = (left + right) / 2
    mergeSort(left: left, right: mid)
    mergeSort(left: mid + 1, right: right)
    
    // merge
    var lIndex = left // 병합시킬 왼쪽 배열의 인덱스
    var rIndex = mid + 1 // 병합시킬 오른쪽 배열의 인덱스
    
    for i in left...right {
      if lIndex > mid { // 왼쪽 배열을 다 병합 했을 경우
        temp[i] = arr[rIndex]
        rIndex += 1
      } else if rIndex > right { // 오른쪽 배열을 다 병합 했을 경우
        temp[i] = arr[lIndex]
        lIndex += 1
      } else if arr[lIndex].power > arr[rIndex].power {
        // 왼쪽 배열의 원소가 오른쪽 배열의 원소보다 값이 클 때
        // 즉 앞서 있는 사람의 실력이 더 좋을 땐 순위의 변동이 없다.
        temp[i] = arr[lIndex]
        lIndex += 1
      } else { // 뒤에 있는 사람이 실력이 좋을 때는 순위의 변동이 있으므로 갱신해줘야 한다.
        temp[i] = arr[rIndex]
        ans[arr[rIndex].index] -= (mid - lIndex + 1)
        rIndex += 1
      } // end of if
    } // end of for
    
    // 임시 배열에 저장해두었던 값 갱신
    for i in left...right {
      arr[i] = temp[i]
    }
  }// end of if
} // end of func

// 입력 부분
for i in 1...n {
  arr[i].index = i
  ans[i] = i
  arr[i].power = readLine()!.split(separator: " ").compactMap{ Int(String($0)) }[0]
}

mergeSort(left: 1, right: n)

for i in 1...n {
  print(ans[i])
}

File: BOJ_Java/BOJ_2517_달리기_P4.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;


public class BJ2517 {
	static class Player { // 문제를 풀기 위해 현재 순위와 실력을 저장할 구조체. 
		int index;
		int power;
	}
	
	static Player[] arr = new Player[500001]; // 정보를 저장할 배열
	static Player[] temp = new Player[500001]; // merge sort를 위해 사용될 임시 배열
	static int[] ans = new int[500001]; // 출력할 답 배열
	
	static void mergeSort(int l, int r) {
		if(l == r) return;
		else {
			int m = (l + r) / 2;
			mergeSort(l, m);
			mergeSort(m + 1, r);
			
			// merge sort의 개념은 왼쪽 배열과 오른쪽을 합병하는 것.
			int lIndex = l; // 배열시킬 왼쪽 배열의 인덱스
			int rIndex = m + 1; // 배열시킬 오른쪽 배열의 인덱스
			for(int i = l; i <= r; i++) {
				if(lIndex > m) temp[i] = arr[rIndex++]; // 왼쪽 배열을 다 병합했을 경우
				else if(rIndex > r) temp[i] = arr[lIndex++]; // 오른쪽 배열을 다 병합했을 경우
				// 앞서 있는 선수가 뒷사람보다 앞서 있는건 볼 필요가 없다. 그대로 저장
				else if(arr[lIndex].power > arr[rIndex].power) temp[i] = arr[lIndex++];
				// 하지만 뒤에 있는 선수가 앞으로 갈때는 앞으로 갈때는 갱신.  
				else {
					temp[i] = arr[rIndex];
					ans[arr[rIndex].index] -= (m - lIndex + 1); // 왼쪽에 배열의 원소의 개수만큼 빼줘서 업데이트 시켜준다.
					rIndex++;
				}
			}
			
			for(int i = l; i <= r; i++) {
				arr[i] = temp[i];
			}
		}
	}
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		int n = Integer.parseInt(br.readLine());
		for(int i = 1; i <= n; i++) {
			arr[i] = new Player();
			ans[i] = arr[i].index = i;
			arr[i].power = Integer.parseInt(br.readLine());
		}
		mergeSort(1, n);
		for(int i = 1; i <= n; i++) bw.write(ans[i] + "\n");
		bw.flush();
	}
}
